### PR TITLE
[PT-D][Tensor Parallelism] Fix parallelize_linear, complete parallelize_module and code refactor

### PR DESCRIPTION
**Pull request by fduwjj**
1. The current rowwise linear is wrong so we need to fix it. The input should be shard(-1)
2. Add _parallelize_linear into parallelize_module
3. Refactor test.

### CODE REVIEW

File: spmd/tensor/__init__.py [hunk 1]
```diff
@@ -168,7 +169,11 @@ def replicate_module_params_buffers(m: nn.Module, mesh: DeviceMesh) -> None:
 
     # register input_fn as module forward pre hook
     if input_fn is not None:
-        module.register_forward_pre_hook(lambda _, inputs: input_fn(inputs, device_mesh))  # type: ignore[misc]
+        dim = kwargs.get("input_shard_dim", None)
```
Comment (wanchaol) [inline spmd/tensor/__init__.py:172]: what's the reason we need this? shouldn't this goes to the make_input_shard/replicate functions directly?

File: spmd/tensor/parallel/api.py [hunk 2]
```diff
@@ -74,6 +74,10 @@ def parallelize_module(  # type: ignore[return]
         device_mesh = _create_1d_device_mesh(device_mesh, tp_mesh_dim)
 
     if isinstance(parallelize_plan, ParallelStyle):
+        # RowwiseParallel or ColwiseParallel
+        if not isinstance(parallelize_plan, PairwiseParallel):
```
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:78]: We should do explicitly check IMO for Rowwise and Colwise

File: spmd/tensor/parallel/api.py [hunk 3]
```diff
@@ -251,6 +254,7 @@ def _parallelize_linear(
             _rowwise_parallelize_linear_fn,
             input_fn=parallel_style._prepare_input,  # type: ignore[arg-type, misc] # pyre-ignore[6]
             output_fn=parallel_style._prepare_output,  # type: ignore[arg-type, misc] # pyre-ignore[6]
+            input_shard_dim=-1,
```
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:257]: I think this should goes into prepare_input of rowwise parallel style directly?

File: spmd/tensor/__init__.py (hunk 1 above)
Comment (fduwjj) [reply spmd/tensor/__init__.py:172]: The function itself is stateless, the dim is not stored there.

File: spmd/tensor/parallel/api.py (hunk 2 above)
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:78]: OK

File: spmd/tensor/parallel/api.py (hunk 3 above)
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:257]: Is there a way for function to store a state?

File: spmd/tensor/__init__.py (hunk 1 above)
Comment (wanchaol) [reply spmd/tensor/__init__.py:172]: I see, but we have a default argument for dim in make_input_shard? should we put the -1 there?

also, maybe we can use functools.partial to generate a partial function that already have the state?

File: spmd/tensor/parallel/api.py (hunk 3 above)
Comment (awgu) [reply spmd/tensor/parallel/api.py:257]: You _could_ store an attribute on the function itself. 
```
>>> def f():
>>>     pass
>>> f.a = 1
>>> f.a
1
```
I am not sure if this is recommended, but there _is_ precedent for this in [old FSDP code](https://github.com/pytorch/pytorch/blob/77d7f2c65945438e0292b270998cea07c0d9d3d8/torch/distributed/fsdp/wrap.py#L226-L227):
```
size_based_auto_wrap_policy.EXCLUDE_WRAP_MODULES = {nn.ModuleList, nn.ModuleDict}  # type: ignore[attr-defined]
size_based_auto_wrap_policy.FORCE_LEAF_MODULES = {nn.MultiheadAttention}  # type: ignore[attr-defined]
```
where `size_based_auto_wrap_policy` is a function.

Comment (fduwjj) [reply spmd/tensor/parallel/api.py:257]: We will create a new prepare_input func instead.

File: spmd/tensor/parallel/style.py [hunk 4]
```diff
@@ -101,6 +101,30 @@ def make_input_shard_1d(
         )
 
 
+def make_input_shard_1d_dim_last(
```
Comment (XilunWu) [inline spmd/tensor/parallel/style.py:104]: thx for fixing the input sharding func

File: spmd/tensor/parallel/style.py [hunk 5]
```diff
@@ -144,14 +168,14 @@ def make_output_shard_1d(
     Convert Output DTensor to a sharded DTensor. This will be used in ParallelStyle.
     Args:
         output (DTensor): output of module to be converted.
-        device_mesh (Optional[DeviceMesh]): :class:``DeviceMesh`` object needed to
+        device_mesh (Optional[DeviceMesh]): :class:`DeviceMesh` object needed to
```
Comment (XilunWu) [inline spmd/tensor/parallel/style.py:171]: (Optional[DeviceMesh]) -> (:class:`DeviceMesh`, optional) 
https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#docstring-type-formatting

File: spmd/tensor/parallel/style.py [hunk 6]
```diff
@@ -165,13 +189,13 @@ def make_output_replicate_1d(
     Convert Output DTensor to a replicated DTensor. This will be used in ParallelStyle.
     Args:
         output (DTensor): output of module to be converted.
-        device_mesh (Optional[DeviceMesh]): :class:``DeviceMesh`` object needed to
+        device_mesh (Optional[DeviceMesh]): :class:`DeviceMesh` object needed to
```
Comment (XilunWu) [inline spmd/tensor/parallel/style.py:192]: ditto

File: spmd/tensor/parallel/style.py [hunk 7]
```diff
@@ -185,13 +209,13 @@ def make_output_tensor(
     Convert Output DTensor to a replicated DTensor first and then convert it to Tensor.
     Args:
         output (DTensor): output of module to be converted.
-        device_mesh (Optional[DeviceMesh]): :class:``DeviceMesh`` object needed to
+        device_mesh (Optional[DeviceMesh]): :class:`DeviceMesh` object needed to
```
Comment (XilunWu) [inline spmd/tensor/parallel/style.py:212]: ditto

File: spmd/tensor/parallel/style.py [hunk 8]
```diff
@@ -185,13 +209,13 @@ def make_output_tensor(
...
             replicate the output and it needs to be a 1D ``device_mesh`` and we will
             throw exceptions if a non-1D ``device_mesh`` is passed in. If no
             ``device_mesh`` is passed in, we will reuse the one from output.
             Default: ``None``
     Return:
-        A :class:``torch.Tensor`` object converted from output DTensor.
+        A :class:`torch.Tensor` object converted from output DTensor.
```
Comment (XilunWu) [inline spmd/tensor/parallel/style.py:218]: DTensor -> :class:`DTensor` maybe? Not strictly necessary however.

File: test/spmd/tensor/parallel/test_parallelize_api.py [hunk 9]
```diff
@@ -79,59 +79,64 @@ def test_creat_1d_device_mesh_error(self):
         ):
             _create_1d_device_mesh(mesh, 3)
 
-    def _check_module(
-        self, local_module, dist_module, inp_size, local_input_fn, device_mesh
+    def _compare_params(
```
Comment (XilunWu) [inline test/spmd/tensor/parallel/test_parallelize_api.py:82]: thx for refactoring this function!

File: spmd/tensor/__init__.py [hunk 10]
```diff
@@ -1,5 +1,5 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
-from typing import Optional, Sequence, Callable, cast
+from typing import Any, Optional, Sequence, Callable, cast
```
Comment (wanchaol) [inline spmd/tensor/__init__.py:2]: nit: remove the Any import to avoid lint failure?

File: spmd/tensor/parallel/api.py [hunk 11]
```diff
@@ -278,19 +284,19 @@ def _parallelize_multihead_attn(
 
     Args:
         module (nn.Module):
```
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:286]: can we change the all the parameter comment, i.e. :class:`nn.Module` inside the "(:class:`nn.Module`) instead? that would match more to the pytorch comment style.

File: test/spmd/tensor/parallel/test_parallelize_api.py [hunk 12]
```diff
@@ -79,59 +79,64 @@ def test_creat_1d_device_mesh_error(self):
...
-                    ).to_local(),
-                )
+        # check backward and ensure gradients are same
+        self._compare_params(local_module, dist_module, rowwise, True)
 
-            local_optim.step()
-            dist_optim.step()
+        local_optim.step()
+        dist_optim.step()
+        self._compare_params(local_module, dist_module, rowwise, True)
```
Comment (XilunWu) [inline test/spmd/tensor/parallel/test_parallelize_api.py:135]: True -> False since we want to check correctness of param update?

**Review by XilunWu — COMMENTED** [review]
Nice work! left some comments.

**Review by wanchaol — APPROVED** [review]
looks good, one suggestion about comment style, thanks for the fix!

File: spmd/tensor/parallel/api.py (hunk 11 above)
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:286]: sure. Let me do that.

File: test/spmd/tensor/parallel/test_parallelize_api.py (hunk 12 above)
Comment (fduwjj) [reply test/spmd/tensor/parallel/test_parallelize_api.py:135]: opps.

File: spmd/tensor/parallel/style.py (hunk 8 above)
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:218]: DTensor part will be removed.
